File: Algorithm_784_Letter Case Permutation.cpp
#include <vector>
#include <string>
using namespace std;
class Solution {
public:
    vector<string> res;
    void addString(string S, int i)
    {
        if (i==S.size())
        {
            res.push_back(S);
            return;
        }
        else if (!isalpha(S[i])) addString(S,i+1);
        else
        {
            addString(S,i+1);
            S[i]=S[i]>='a' && S[i]<='z'?char(S[i]-'a'+'A'):char(S[i]-'A'+'a');
            addString(S,i+1);
        }
    }
    vector<string> letterCasePermutation(string S) {
        addString(S,0);
        return res;
    }
};

File: Algorithm_784_Letter Case Permutation.java
import java.util.*;

// 直接用回溯法，对digit，直接append然后递归
// 对letter，分别转换成大写和小写，然后递归

class Solution {
    private List<String> result;
    public List<String> letterCasePermutation(String s) {
        this.result = new ArrayList<>();
        backtracking(s.toCharArray(), 0, "");
        return this.result;
    }
    private void backtracking(char[] s, int idx, String cur) {
        if (idx == s.length) {
            this.result.add(cur);
            return;
        }
        if (Character.isDigit(s[idx])) {
            backtracking(s, idx + 1, cur + s[idx]);
        } else {
            backtracking(s, idx + 1, cur + Character.toLowerCase(s[idx]));
            backtracking(s, idx + 1, cur + Character.toUpperCase(s[idx]));
        }
    }
}